packages/nitro-protocol/contracts/examples/EmbeddedApplication.sol

```diff
@@ -0,0 +1,438 @@
+        }
+
+        // validate the supplied support proof
+        // extract the allocation
+        Outcome.AllocationItem[] memory Xallocation = validateSupportProofForX(
+            fromAppData,
+            toAppData
+        );
+
+        // ensure A,B part of the outcome of X has been absorbed into the outcome of J
+        require(
+            Xallocation[0].amount == toAllocation[0].amount &&
+                Xallocation[1].amount == toAllocation[1].amount &&
+                Xallocation[0].destination == toAllocation[0].destination &&
+                Xallocation[1].destination == toAllocation[1].destination,
+            'X / J outcome mismatch'
+        );
+        return true;
+    }
+
+    function validateSupportProofForX(AppData memory fromAppData, AppData memory toAppData)
+        internal
+        pure
+        returns (Outcome.AllocationItem[] memory Xallocation)
+    {
+        // The following checks follow the protocol-level validTransition function
+        // They are the members of FixedPart that do not affect the channelId
+        require(
+            fromAppData.supportProofForX.fixedPart.appDefinition ==
+                toAppData.supportProofForX.fixedPart.appDefinition,
+            'X.appDefinition changed'
+        );
+        require(
+            fromAppData.supportProofForX.fixedPart.challengeDuration ==
+                toAppData.supportProofForX.fixedPart.challengeDuration,
+            'X.challengeDuration changed'
+        ); // this is actually never used so could be 0
+
+        // A support proof requires 2 signatures
+        // But it may have either 1 or two states.
+        // If both signatures on the same state, it is supported
+        // Else one signature per state and we must check for a valid transition
+        require(
+            toAppData.supportProofForX.variableParts.length == 1 ||
+                toAppData.supportProofForX.variableParts.length == 2,
+            '1 or 2 states required'
+        );
+
+        bytes32 appPartHash = keccak256(
+            abi.encode(
+                toAppData.supportProofForX.fixedPart.challengeDuration,
+                toAppData.supportProofForX.fixedPart.appDefinition,
+                toAppData.supportProofForX.variableParts[0].appData
+            )
+        );
+
+        // hash the greatest state first (either the later of a pair, or the only state provided)
+        bytes32 greaterStateHash = keccak256(
+            abi.encode(
+                IForceMove.State(
+                    toAppData.supportProofForX.turnNumTo,
+                    false, // Assume isFinal is false
+                    toAppData.channelIdForX,
+                    appPartHash,
+                    keccak256(
+                        toAppData.supportProofForX.variableParts[toAppData
+                            .supportProofForX
+                            .variableParts
+                            .length - 1]
+                            .outcome
```

Comment:
Readability here might benefit from defining a `greaterStateOutcome` variable to avoid these line breaks

Reply:
maybe define
```
uint256 greaterStateTurnNum = toAppData.supportProofForX.variableParts.length - 1
```
?

Reply:
6019e82